tests/components/rest/test_binary_sensor.py

```diff
@@ -1,267 +1,381 @@
+                "username": "my username",
+                "password": "my password",
+                "headers": {"Accept": CONTENT_TYPE_JSON},
+            }
+        },
+    )
+
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+
+@respx.mock
+async def test_setup_post(hass):
+    """Test setup with valid configuration."""
+    respx.post("http://localhost", status_code=200, content="{}")
+    assert await async_setup_component(
+        hass,
+        "binary_sensor",
+        {
+            "binary_sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "POST",
+                "value_template": "{{ value_json.key }}",
+                "payload": '{ "device": "toaster"}',
+                "name": "foo",
+                "verify_ssl": "true",
+                "timeout": 30,
+                "authentication": "basic",
+                "username": "my username",
+                "password": "my password",
+                "headers": {"Accept": CONTENT_TYPE_JSON},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+
+@respx.mock
+async def test_setup_get_off(hass):
+    """Test setup with valid off configuration."""
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": "text/json"},
+        content='{"dog": false}',
+    )
+    assert await async_setup_component(
+        hass,
+        "binary_sensor",
+        {
+            "binary_sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.dog }}",
+                "name": "foo",
+                "verify_ssl": "true",
+                "timeout": 30,
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("binary_sensor.foo")
+    assert state.state == STATE_OFF
+
+
+@respx.mock
+async def test_setup_get_on(hass):
+    """Test setup with valid on configuration."""
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": "text/json"},
+        content='{"dog": true}',
+    )
+    assert await async_setup_component(
+        hass,
+        "binary_sensor",
+        {
+            "binary_sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.dog }}",
+                "name": "foo",
+                "verify_ssl": "true",
+                "timeout": 30,
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("binary_sensor.foo")
+    assert state.state == STATE_ON
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)